frontend/src/utils/timerange.js

```diff
@@ -0,0 +1,197 @@
+import { DateTime } from "luxon";
+
+const NANOS_PER_SECOND = 1_000_000_000n;
+const TIMERANGE_REGEX =
+  /^(?<startInclusive>\[|\()?(?:(?<startSeconds>-?(?:0|[1-9][0-9]*)):(?<startNanos>0|[1-9][0-9]{0,8}))?(?:_(?:(?<endSeconds>-?(?:0|[1-9][0-9]*)):(?<endNanos>0|[1-9][0-9]{0,8}))?)?(?<endInclusive>\]|\))?$/;
+const INFINITE_PATTERNS = new Set([
+  "_",
+  "(_)",
+  "[_]",
+  "(_]",
+  "[_)",
+  "(_",
+  "[_",
+  "_)",
+  "_]",
+]);
+const EMPTY_PATTERNS = new Set([
+  "",
+  "()",
+  "[]",
+  "(]",
+  "[)",
+  ")",
+  "]",
+  "(",
+  "[",
+]);
+
+const isInclusive = (bracket) => bracket !== "(";
+const isEndInclusive = (bracket) => bracket !== ")";
+
+const createTimerangeResult = (includesStart, start, end, includesEnd) => ({
+  includesStart,
+  start,
+  end,
+  includesEnd,
+});
+
+/**
+ * Converts seconds and nanoseconds to a BigInt representing total nanoseconds
+ *
+ * @param {string} seconds - Seconds from epoch (string needed to handle negative values correctly)
+ * @param {number} nanos - Nanoseconds component
+ * @returns {BigInt} Total nanoseconds as BigInt
+ */
+const secondsNanosToBigInt = (secondsStr, nanos) => {
+  const isNegative = secondsStr?.toString().startsWith("-");
+  const absSeconds = Math.abs(parseInt(secondsStr || "0", 10));
+  const absTotalNanoSeconds =
+    BigInt(absSeconds) * NANOS_PER_SECOND + BigInt(nanos);
+  return isNegative ? -absTotalNanoSeconds : absTotalNanoSeconds;
+};
+
+/**
+ * Parses a timerange string into an object with BigInt values
+ *
+ * @param {string} timerange - The timerange string to parse
+ * @returns {Object} Object with start and end as BigInt nanoseconds
+ * @throws {Error} If the timerange string format is invalid
+ */
+export const parseTimerange = (timerange) => {
+  // Handle special case: infinite timerange
+  if (INFINITE_PATTERNS.has(timerange)) {
+    return createTimerangeResult(true, null, null, true);
+  }
+
+  // Handle special case: empty timerange
+  if (EMPTY_PATTERNS.has(timerange)) {
+    return createTimerangeResult(false, 0n, 0n, false);
+  }
+
+  // Regular expression to match timerange components
+  const match = timerange.match(TIMERANGE_REGEX);
+  if (!match) {
+    throw new Error("Invalid timerange string format");
```

Comment:
is this error being caught further up - I think if you throw an error you could cause the app to unmount if the error bubbles

Reply:
Yep, fair point. I've removed the throwing of errors throughout and added try/catch. The functions will now return default empty timeranges if errors occur so that the components using these functions will not receive "bad data".